### PR TITLE
feat(framework): Add NestJS `serve` handler

### LINKED ISSUE #6552: 📚 Docs Feedback: Add informations about how to use Novu with NestJs
### Duplicates

- [X] I have searched the existing issues

### Related page

https://docs.nestjs.com/middleware

### Kind of issue

Missing information

### Issue description

NestJs is like a superset of express, therefore a user can leverage the middleware implementation similar to express.

Afterwards you find some of our implemented code to use for your documentation:

## the express middleware, using require, because we are using commonjs
``` novu.middleware.ts

import { Injectable, NestMiddleware } from '@nestjs/common';
import { Request, Response, NextFunction } from 'express';
import { serve } from '@novu/framework/dist/servers/express';
import workflows from './workflows';

@Injectable()
export class NovuMiddleware implements NestMiddleware {
    use(req: Request, res: Response, next: NextFunction) {
        // Using require instead of import syntax, because novu export is not commonjs
        require("@novu/framework/express").serve({
            workflows,
        } as Parameters<typeof serve>['0'])(req, res, next)
    }
}
```

## Probably the right implementation
``` novu.middleware.ts

import { Injectable, NestMiddleware } from '@nestjs/common';
import { Request, Response, NextFunction } from 'express';
import { serve } from '@novu/framework/express';
import workflows from './workflows';

@Injectable()
export class NovuMiddleware implements NestMiddleware {
    use(req: Request, res: Response, next: NextFunction) {
        serve({
            workflows,
        })(req, res, next)
    }
}
```

## app.module.ts
``` app.module.ts

import { MiddlewareConsumer, Module } from '@nestjs/common';
import { AppController } from './app.controller';
import { AppService } from './app.service';
import { NovuMiddleware } from './shared/novu/novu.middleware';

@Module({
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule {
  configure(consumer: MiddlewareConsumer) {
    consumer
      .apply(NovuMiddleware)
      .forRoutes('/api/novu');
  }
}
```

### Context 🔦

_No response_

### PR DESCRIPTION
**Pull request by rifont**
### What changed? Why was the change needed?

* Add idiomatic NestJS `serve` handler using NestJS modules and controllers
* Add NestJS playground application to demonstrate best practice usage of the new handler

closes https://github.com/novuhq/novu/issues/6552

### Screenshots

_Basic DX (without NestJS dependency injection, only basic content Workflows using `NovuModule.register`):_
```typescript
import { Module } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { workflow } from '@novu/framework';
import { NovuModule } from '@novu/framework/nest';

@Module({
  imports: [
    ConfigModule.forRoot({
      envFilePath: '.env',
    }),
    NovuModule.register({
      apiPath: '/api/novu',
      workflows: [
        workflow('welcome-workflow', async ({ step, payload }) => {
          await step.email('send-email', async (controls) => ({
            subject: `${controls.greeting}`,
            body: `We are glad you are here! UserId: ${payload.userId}`,
          }));
        }),
      ],
    }),
  ],
})
export class AppModule {}
```

_Advanced DX (with NestJS dependency injection inside workflows using `NovuModule.registerAsync`):_

```typescript
// #############################################
// ########### notification.service.ts #########
// #############################################
// demonstrating declaration of workflows with access to injected services

import { Injectable } from '@nestjs/common';
import { workflow } from '@novu/framework';
import { z } from 'zod';
import { UserService } from './user.service';

@Injectable()
export class NotificationService {
  constructor(private readonly userService: UserService) {}

  public welcomeWorkflow() {
    return workflow(
      'welcome-workflow',
      async ({ step, payload }) => {
        await step.email(
          'send-email',
          async (controls) => {
            const user = this.userService.getUser(payload.userId);

            return {
              subject: `${controls.greeting}, ${user.name}`,
              body: `We are glad you are here! Email: ${user.email}`,
            };
          },
          {
            controlSchema: z.object({
              greeting: z.string().default('Welcome to our platform'),
            }),
          },
        );
      },
      {
        payloadSchema: z.object({
          userId: z.string(),
        }),
      },
    );
  }
}



// ############################################
// ############# app.module.ts ################
// ############################################
// demonstrating module initialization

import { Module } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { NovuModule } from '@novu/framework/nest';
import { AppController } from './app.controller';
import { AppService } from './app.service';
import { NotificationService } from './notification.service';
import { UserService } from './user.service';

@Module({
  imports: [
    /*
     * IMPORTANT: ConfigModule must be imported before NovuModule to ensure
     * environment variables are loaded before the NovuModule is initialized.
     *
     * This ensures that NOVU_SECRET_KEY is available when the NovuModule is initialized.
     */
    ConfigModule.forRoot({
      envFilePath: '.env',
    }),
    NovuModule.registerAsync({
      imports: [AppModule],
      useFactory: (notificationService: NotificationService) => ({
        apiPath: '/api/novu',
        workflows: [notificationService.welcomeWorkflow()],
      }),
      inject: [NotificationService],
    }),
  ],
  controllers: [AppController],
  providers: [AppService, NotificationService, UserService],
  exports: [NotificationService],
})
export class AppModule {}



// #############################################
// ############# app.controller.ts #############
// #############################################
// demonstrating type-safe `workflow().trigger` usage
import { Controller, Get, Param } from '@nestjs/common';
import { AppService } from './app.service';
import { NotificationService } from './notification.service';
import { UserService } from './user.service';

@Controller()
export class AppController {
  constructor(
    private readonly appService: AppService,
    private readonly notificationService: NotificationService,
    private readonly userService: UserService,
  ) {}

  @Get()
  getHello(): string {
    return this.appService.getHello();
  }

  @Get('/welcome/:userId')
  public async sendWelcomeNotification(@Param('userId') userId: string) {
    const user = this.userService.getUser(userId);

    return this.notificationService.welcomeWorkflow().trigger({
      payload: { userId },
      to: {
        subscriberId: userId,
        email: user.email,
      },
    });
  }
}

```

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Special notes for your reviewer

</details>

### CODE REVIEW

File: packages/framework/src/handler.ts [hunk 1]
```diff
@@ -132,7 +132,7 @@ export class NovuRequestHandler<Input extends any[] = any[], Output = any> {
   private async handleAction({ actions }: { actions: HandlerResponse<Output> }): Promise<IActionResponse> {
     const url = await actions.url();
     const method = await actions.method();
-    const action = url.searchParams.get(HttpQueryKeysEnum.ACTION) || GetActionEnum.HEALTH_CHECK;
+    const action = url.searchParams.get(HttpQueryKeysEnum.ACTION) || '';
```
Comment (rifont) [inline packages/framework/src/handler.ts]:
```suggestion
    const action = url.searchParams.get(HttpQueryKeysEnum.ACTION) || GetActionEnum.HEALTH_CHECK;
```

File: packages/framework/src/servers/nest/nest.module-definition.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { ConfigurableModuleBuilder } from '@nestjs/common';
+import { NovuModuleOptions } from './nest.interface';
+
+// use ConfigurableModuleBuilder, because building dynamic modules from scratch is painful
```
Comment (rifont) [inline packages/framework/src/servers/nest/nest.module-definition.ts:4]: Following best practices to use ConfigurableModuleBuilder per [Dynamic Modules docs](https://docs.nestjs.com/fundamentals/dynamic-modules#configurable-module-builder)

File: packages/framework/tsconfig.json [hunk 3]
```diff
@@ -13,6 +13,7 @@
     "sourceMap": true,
     "rootDir": ".",
     "outDir": "./dist",
+    "experimentalDecorators": true,
```
Comment (rifont) [inline packages/framework/tsconfig.json:16]: Necessary to use NestJS decorators.

**Review by SokratisVidros — APPROVED** [review]
😍

**Review by SokratisVidros — APPROVED** [review]

**Comment by paulwer** [comment]
While using the existing SDK without the distinct novu nestjs integration, we have leveraged the batch endpoint quite often to reduce the amount of outgoing requests. Therefore it would be awesome to have like Workflow Types which can be eigher reused within a batch function or a typesafe implementation choosing the right types based on the workflow name within the batch fn. 

Besides this, this is a great feature and implementation, thanks for the work❤️

**Comment by rifont** [comment]
> While using the existing SDK without the distinct novu nestjs integration, we have leveraged the batch endpoint quite often to reduce the amount of outgoing requests. Therefore it would be awesome to have like Workflow Types which can be eigher reused within a batch function or a typesafe implementation choosing the right types based on the workflow name within the batch fn.
> 
> Besides this, this is a great feature and implementation, thanks for the work❤️

Thanks for sharing the use-case. We'll consider bulk type-safe triggering separately from the NestJS Framework handler, we will likely export a utility helper from `@novu/framework` rather than Web-Framework specific that does the following:
1. Aggregates all the defined `workflow(...)` functions for usage in the serve handlers
2. Performs Typescript generic manipulation of the defined `workflow(...)` functions with a discriminated union on the `workflowId` to derive the necessary types for triggering.

I'm raising this for consideration in our backlog 🙏

**Comment by paulwer** [comment]
Thanks for the clarification. Great progress.

**Pull request merged**